### PR TITLE
[CPU] [ARM] JIT GELU Erf

### PR DESCRIPTION
**Pull request by eshoguli**
### Details:
 - *[CPU] [AARCH64] jit gelu erf*

### Tickets:
 - *CVS-138192*

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@xuchen-intel Please take a look

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 1]
```diff
@@ -391,6 +391,144 @@ std::set<std::vector<element::Type>> jit_exp_emitter::get_supported_precisions(c
...
+    }
+
+    using TReg = typename dnnl::impl::cpu::aarch64::cpu_isa_traits<isa>::TReg;
+    const TReg vmm_src(in_vec_idxs[0]);
+    const TReg vmm_dst(out_vec_idxs[0]);
+
+    const TReg vmm_aux0(aux_vec_idxs[0]);
+    const TReg vmm_aux1(aux_vec_idxs[1]);
+    const TReg vmm_aux2(aux_vec_idxs[2]);
+    const TReg vmm_aux3(aux_vec_idxs[3]);
```
Comment (xuchen-intel) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:441]: If the return value of exp_emitter->get_aux_gprs_count() is independent the implementation of jit_gelu_erf_emitter, then perhaps we are not sure aux_vec_idxs[3] is always valid. If the rest of gelu computation(exclude exp) has constraints on the number of aux_vecs, we should update that in jit_gelu_erf_emitter::get_aux_vecs_count().

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 2]
```diff
@@ -391,6 +391,144 @@ std::set<std::vector<element::Type>> jit_exp_emitter::get_supported_precisions(c
...
+                                         dnnl::impl::cpu::aarch64::cpu_isa_t host_isa,
+                                         const ov::element::Type exec_prc) : jit_emitter(host, host_isa, exec_prc) {
+    prepare_table();
+    exp_emitter = std::make_unique<jit_exp_emitter>(h, host_isa, exec_prc);
+}
+
+size_t jit_gelu_erf_emitter::get_inputs_count() const { return 1; }
+
+size_t jit_gelu_erf_emitter::get_aux_vecs_count() const {
+    return exp_emitter->get_aux_vecs_count() + 3;
```
Comment (xuchen-intel) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:413]: If the return value of exp_emitter->get_aux_gprs_count() is independent from the implementation of jit_gelu_erf_emitter, then perhaps we'd better use sth. like:
`return std:max(exp_emitter->get_aux_vecs_count() + k1, k2);`
When exp_emitter->get_aux_vecs_count() is large, we reuse its aux_vecs for the rest of gelu computation. When exp_emitter->get_aux_vecs_count() is small, k2 can guarantee enough aux_vecs for the rest of gelu computation.

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 3]
```diff
@@ -391,6 +391,144 @@ std::set<std::vector<element::Type>> jit_exp_emitter::get_supported_precisions(c
...
+void jit_gelu_erf_emitter::emit_impl(const std::vector<size_t> &in_vec_idxs, const std::vector<size_t> &out_vec_idxs) const {
+    if (host_isa_ == dnnl::impl::cpu::aarch64::asimd) {
+        emit_isa<dnnl::impl::cpu::aarch64::asimd>(in_vec_idxs, out_vec_idxs);
+    } else {
+        OV_CPU_JIT_EMITTER_THROW("Can't create jit eltwise kernel");
+    }
+}
+
+template <dnnl::impl::cpu::aarch64::cpu_isa_t isa>
+void jit_gelu_erf_emitter::emit_isa(const std::vector<size_t> &in_vec_idxs, const std::vector<size_t> &out_vec_idxs) const {
```
Comment (xuchen-intel) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:429]: Just for my understanding. Could you please share the reference material about the formular of this implementation? Thanks!

**Review by xuchen-intel — CHANGES_REQUESTED** [review]

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp (hunk 1 above)
Comment (eshoguli) [reply src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:441]: Minimum amount of auxiliary registers is 7. It's defined in `jit_gelu_erf_emitter::get_aux_vecs_count()`. It's safe to use.

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 4]
```diff
@@ -391,6 +391,142 @@ std::set<std::vector<element::Type>> jit_exp_emitter::get_supported_precisions(c
...
+                                           dnnl::impl::cpu::aarch64::cpu_isa_t host_isa,
+                                           const ov::element::Type exec_prc) : jit_emitter(host, host_isa, exec_prc) {
+    prepare_table();
+    exp_emitter = std::make_unique<jit_exp_emitter>(h, host_isa, exec_prc);
+}
+
+size_t jit_gelu_erf_emitter::get_inputs_count() const { return 1; }
+
+size_t jit_gelu_erf_emitter::get_aux_vecs_count() const {
+    return std::max<size_t>(exp_emitter->get_aux_vecs_count() + 3, 7);
```
Comment (xuchen-intel) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:413]: Maybe we can save 2 aux_vecs. Here are my thoughts:
1. As vmm_dst is used very late, vmm_aux can reuse vmm_dst.
2. If we delay the computation of t to when we compute -exp(-x*x)*t, then vmm_aux_t can reuse vmm_aux0.

What do you think? If they make sense, I hope they won't take much efferts, otherwise we can do it in separate PR.

**Review by xuchen-intel — CHANGES_REQUESTED** [review]

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 5]
```diff
@@ -391,6 +391,142 @@ std::set<std::vector<element::Type>> jit_exp_emitter::get_supported_precisions(c
...
+            { vmm_aux.getIdx() },
+            { vmm_aux_dst.getIdx() },
+            aux_vec_idxs,
+            aux_gpr_idxs);
+    h->ld1r(vmm_aux2.s, table_val2("sign_mask"));
+    // vmm_aux_dst = -exp(-x*x)
+    h->orr(vmm_aux_dst.b16, vmm_aux_dst.b16, vmm_aux2.b16);
+
+    // get sign
+    h->ld1r(vmm_aux2.s, table_val2("sign_mask"));
```
[the remark this answers is not in the document]
Comment (eshoguli) [reply src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:473]: removed, thanks

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp (hunk 4 above)
Comment (eshoguli) [reply src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:413]: Thanks for comment, extra auxiliary registers are potential issue in the future. But your suggestions are not clear, can you provide more details? Thanks!

Note, please, source and destination registers can be the same. If not, then value in source register can not be changed. 

Some comments:
1. It's not possible to use destination register if source value should be stored.
2. It doesn't matter what value we should store in auxillary register which are not used by `exp_emitter`: `x / sqrt(2)` or `t`. We need to store this value in auxillary register which are not used `exp_emitter` (`register index >= exp_emitter->get_aux_vecs_count()`).

Comment (xuchen-intel) [reply src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:413]: 1. Cool! Thanks for the reminder. For sure, jit_gelu_erf_emitter itself also supports shared src and dst. Let do not use vmm_dst.
2. I mean can we move the computation of `t` https://github.com/eshoguli/openvino/blob/es/cpu/jit_gelu_erf/src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp#L451-L456 to right before the computation of `-exp(-x*x)*t` https://github.com/eshoguli/openvino/blob/es/cpu/jit_gelu_erf/src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp#L475 . Then we don't need an extra aux_vec for vmm_aux_t. Since exp is done, and its 4 aux_vecs are free. Perhaps we can reuse vmm_aux3 (Previously I was suggesting vmm_aux0, but it seems would be ocuppied). What do you think?

Comment (eshoguli) [reply src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:501]: Got it, thanks. If we move `t` calculation after `exp_emitter->emit_code`, then we need to store `x` value , which is stored in `vmm_aux0`. If not, I need more details about your suggestion. Can we do offline? it in separate PR?

P.S. `x = abs(x)`, do you remember removed comment `// vmm_aux0 lost sign but it doesn't matter`? :)

Comment (xuchen-intel) [reply src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:501]: You are right! Now I believe all the aux_vecs are necessary. Thanks Edward for the elaboration!

**Review by xuchen-intel — APPROVED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]
